File: Source/Private/Oa/Ml/Rl/Evaluation.cpp
#include <Oa/Ml/Rl/Evaluation.h>

#include <Oa/Core/FnMatrix.h>
#include <Oa/Ml/Autograd.h>
#include <Oa/Ml/Rl/Policy.h>
#include <Oa/Ml/Rl/Rollout.h>

#include "EnvironmentExecution.h"

#include <algorithm>
#include <limits>

OaResult<OaRlEvaluationMetrics> OaFnRl::EvaluateCategorical(
	OaRlEnvironment& InEnvironment,
	OaRlActorCritic& InModel,
	const OaRlEvaluationConfig& InConfig) {
	if (InConfig.Horizon == 0 || InEnvironment.Environments() == 0) {
		return OaStatus::InvalidArgument(
			"OaFnRl::EvaluateCategorical expects a non-empty horizon and environment");
	}
	OA_RETURN_IF_ERROR(InEnvironment.Spec().ValidateDefinition());
	if (InEnvironment.Spec().Action.Kind != OaRlSpaceKind::Discrete) {
		return OaStatus::InvalidArgument(
			"OaFnRl::EvaluateCategorical requires a discrete action space");
	}
	const OaU32 environments = InEnvironment.Environments();
	OaRlRolloutBuffer rollout;
	const OaI64 observationElements =
		InEnvironment.Spec().Observation.ElementsPerEnvironment();
	const OaStatus recorded = InEnvironment.RecordCommands([&]() -> OaStatus {
		auto rolloutResult = OaRlRolloutBuffer::Create({
			.Time = InConfig.Horizon,
			.Environments = environments,
			.ObservationShape = InEnvironment.Spec().Observation.Shape,
		});
		if (rolloutResult.IsError()) return rolloutResult.GetStatus();
		rollout = OaStdMove(*rolloutResult);
		rollout.Reset();
		OA_RETURN_IF_ERROR(InEnvironment.ResetEnvironment(InConfig.Seed));
		OaGradNo noGrad;
		for (OaU32 step = 0; step < InConfig.Horizon; ++step) {
			const OaMatrix observation = InEnvironment.Observation();
			const OaMatrix flat = OaFnMatrix::Reshape(observation,
				{static_cast<OaI64>(environments), observationElements});
			const OaRlActorCriticOutput network = InModel.Evaluate(flat);
			if (!network.IsValid()) return OaStatus::Error(
				OaStatusCode::FailedPrecondition,
				"OaFnRl::EvaluateCategorical actor/critic evaluation failed");
			const OaMatrix action = OaFnMatrix::SampleLogits(
				network.Logits, 0.0F, 0, 1.0F, InConfig.Seed);
			const OaRlPolicyResult policy = OaFnRl::EvaluateCategoricalPolicy(
				network.Logits, action, network.Value);
			if (!policy.IsValid()) return OaStatus::Error(
				OaStatusCode::FailedPrecondition,
				"OaFnRl::EvaluateCategorical policy evaluation failed");
			auto transition = InEnvironment.StepEnvironment(action);
			if (transition.IsError()) return transition.GetStatus();
			OA_RETURN_IF_ERROR(rollout.Append({
				.Observation = transition->Observation,
				.Action = action,
				.Reward = transition->Reward,
				.Value = network.Value,
				.NextValue = network.Value,
				.LogProbability = policy.LogProbability,
				.Terminated = transition->Terminated,
				.Truncated = transition->Truncated,
			}));
			OA_RETURN_IF_ERROR(InEnvironment.ResetCompleted());
		}
		return OaStatus::Ok();
	});
	if (recorded.IsError()) return recorded;
	auto completion = InEnvironment.Submit();
	if (completion.IsError()) return completion.GetStatus();
	OA_RETURN_IF_ERROR(InEnvironment.Wait(*completion));
	const OaU64 transitions = static_cast<OaU64>(InConfig.Horizon) * environments;
	OaVec<OaF32> reward(static_cast<OaUsize>(transitions));
	OaVec<OaU8> terminated(static_cast<OaUsize>(transitions));
	OaVec<OaU8> truncated(static_cast<OaUsize>(transitions));
	{
		// The environment owns the execution context that produced the rollout.
		// Keep its engine selected for every host readback instead of falling
		// through to an unrelated ambient compatibility context.
		OaRlEnvironmentRecordingScope scope(InEnvironment);
		OA_RETURN_IF_ERROR(OaFnMatrix::CopyToHost(
			rollout.Batch().Reward, reward.data(),
			transitions * sizeof(OaF32)));
		OA_RETURN_IF_ERROR(OaFnMatrix::CopyToHost(
			rollout.Batch().Terminated, terminated.data(), transitions));
		OA_RETURN_IF_ERROR(OaFnMatrix::CopyToHost(
			rollout.Batch().Truncated, truncated.data(), transitions));
	}

	OaVec<OaF32> episodeReturn(environments, 0.0F);
	OaF64 sum = 0.0;
	OaF32 minimum = std::numeric_limits<OaF32>::infinity();
	OaF32 maximum = -std::numeric_limits<OaF32>::infinity();
	OaU64 completed = 0;
	for (OaU32 step = 0; step < InConfig.Horizon; ++step) {
		for (OaU32 lane = 0; lane < environments; ++lane) {
			const OaUsize index = static_cast<OaUsize>(step) * environments + lane;
			episodeReturn[lane] += reward[index];
			if (terminated[index] != 0 || truncated[index] != 0) {
				sum += episodeReturn[lane];
				minimum = std::min(minimum, episodeReturn[lane]);
				maximum = std::max(maximum, episodeReturn[lane]);
				episodeReturn[lane] = 0.0F;
				++completed;
			}
		}
	}
	return OaRlEvaluationMetrics{
		.EnvironmentSteps = InConfig.Horizon,
		.Transitions = transitions,
		.CompletedEpisodes = completed,
		.MeanCompletedReturn = completed == 0
			? 0.0F : static_cast<OaF32>(sum / static_cast<OaF64>(completed)),
		.MinimumCompletedReturn = completed == 0 ? 0.0F : minimum,
		.MaximumCompletedReturn = completed == 0 ? 0.0F : maximum,
	};
}

File: src/rs/ml/evaluation.rs
//! Explicit policy-evaluation telemetry boundaries.

use crate::{Error, Result, matrix};

use super::{
	ActorCritic, EnvironmentSpaceKind, RolloutBuffer, RolloutConfig, RolloutTransition,
	environment::Environment, policy,
};

/// Fixed deterministic categorical evaluation horizon.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct PolicyEvaluationConfig {
	/// Number of vector-environment steps to evaluate.
	pub horizon: usize,
	/// Reset seed and deterministic greedy operation seed.
	pub seed: u64,
}

impl Default for PolicyEvaluationConfig {
	fn default() -> Self {
		Self {
			horizon: 1000,
			seed: 1,
		}
	}
}

/// Completed categorical evaluation telemetry.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct PolicyEvaluationMetrics {
	/// Vector-environment steps evaluated.
	pub environment_steps: u64,
	/// Individual lane transitions evaluated.
	pub transitions: u64,
	/// Terminated or truncated episodes completed inside the horizon.
	pub completed_episodes: u64,
	/// Mean return across completed episodes, or zero when none completed.
	pub mean_completed_return: f32,
	/// Minimum completed return, or zero when none completed.
	pub minimum_completed_return: f32,
	/// Maximum completed return, or zero when none completed.
	pub maximum_completed_return: f32,
}

/// Deterministically evaluate a categorical policy and return compact telemetry.
///
/// The complete horizon is recorded and submitted once, followed by one exact
/// wait and three compact rollout readbacks. Greedy action selection uses the
/// donor temperature-zero route. The environment remains open and idle.
///
/// # Errors
///
/// Returns an error for zero horizon/lanes, a non-discrete environment, model
/// ownership mismatch, environment/model/policy/rollout recording failure,
/// submission/wait failure, readback failure, or counter overflow.
pub fn evaluate_categorical(
	environment: &mut dyn Environment,
	model: &dyn ActorCritic,
	config: PolicyEvaluationConfig,
) -> Result<PolicyEvaluationMetrics> {
	if config.horizon == 0 || environment.environments() == 0 {
		return Err(Error::invalid_argument(
			"categorical evaluation requires a nonempty horizon and vector environment",
		));
	}
	if environment.spec().action().kind() != EnvironmentSpaceKind::Discrete {
		return Err(Error::invalid_argument(
			"categorical evaluation requires a discrete action space",
		));
	}
	for parameter in model.all_parameters()? {
		if !environment.engine().owns_matrix(&parameter.data()) {
			return Err(Error::invalid_argument(
				"evaluation model must belong to the environment Engine",
			));
		}
	}
	let environments = environment.environments() as usize;
	let observation_shape = environment.spec().observation().shape().to_vec();
	let observation_elements = environment.spec().observation().elements_per_environment();
	let mut rollout = RolloutBuffer::new(
		environment.engine(),
		RolloutConfig {
			time: config.horizon,
			environments,
			observation_shape,
		},
	)?;

	if let Err(error) = record_categorical_evaluation(
		environment,
		model,
		config,
		environments,
		observation_elements,
		&mut rollout,
	) {
		let _ = environment.cancel();
		rollout.abort_unsubmitted();
		return Err(error);
	}
	let event = environment.submit()?;
	environment.wait(&event)?;
	let reward = rollout.batch().reward().read_f32()?;
	let terminated = rollout.batch().terminated().read::<u8>()?;
	let truncated = rollout.batch().truncated().read::<u8>()?;
	summarize(
		config.horizon,
		environments,
		&reward,
		&terminated,
		&truncated,
	)
}

fn record_categorical_evaluation(
	environment: &mut dyn Environment,
	model: &dyn ActorCritic,
	config: PolicyEvaluationConfig,
	environments: usize,
	observation_elements: usize,
	rollout: &mut RolloutBuffer,
) -> Result<()> {
	environment.reset(config.seed)?;
	rollout.reset()?;
	for _ in 0..config.horizon {
		let observation = environment.observation().clone();
		let flat = matrix::reshape(&observation, [environments, observation_elements])?;
		let network = model.evaluate(&flat)?;
		let action = matrix::sample_logits(&network.logits, 0.0, 0, 1.0, config.seed)?;
		let selected = policy::evaluate_categorical(&network.logits, &action, &network.value)?;
		let transition = environment.step(&action)?;
		rollout.append(&RolloutTransition::new(
			transition.observation().clone(),
			action,
			transition.reward().clone(),
			network.value.clone(),
			network.value,
			selected.log_probability,
			transition.terminated().clone(),
			transition.truncated().clone(),
		))?;
		environment.reset_completed()?;
	}
	Ok(())
}

fn summarize(
	horizon: usize,
	environments: usize,
	reward: &[f32],
	terminated: &[u8],
	truncated: &[u8],
) -> Result<PolicyEvaluationMetrics> {
	let transitions = horizon
		.checked_mul(environments)
		.ok_or_else(|| Error::resource_exhausted("evaluation transition count overflows"))?;
	if reward.len() != transitions
		|| terminated.len() != transitions
		|| truncated.len() != transitions
	{
		return Err(Error::failed_precondition(
			"evaluation rollout readback length is inconsistent",
		));
	}
	let mut episode_return = vec![0.0_f32; environments];
	let mut sum = 0.0_f64;
	let mut minimum = f32::INFINITY;
	let mut maximum = f32::NEG_INFINITY;
	let mut completed = 0_u64;
	for step in 0..horizon {
		for (lane, lane_return) in episode_return.iter_mut().enumerate() {
			let index = step * environments + lane;
			*lane_return += reward[index];
			if terminated[index] != 0 || truncated[index] != 0 {
				sum += f64::from(*lane_return);
				minimum = minimum.min(*lane_return);
				maximum = maximum.max(*lane_return);
				*lane_return = 0.0;
				completed = completed
					.checked_add(1)
					.ok_or_else(|| Error::resource_exhausted("completed episode count exhausted"))?;
			}
		}
	}
	Ok(PolicyEvaluationMetrics {
		environment_steps: u64::try_from(horizon)
			.map_err(|_| Error::resource_exhausted("evaluation horizon exceeds u64"))?,
		transitions: u64::try_from(transitions)
			.map_err(|_| Error::resource_exhausted("evaluation transitions exceed u64"))?,
		completed_episodes: completed,
		mean_completed_return: if completed == 0 {
			0.0
		} else {
			(sum / completed as f64) as f32
		},
		minimum_completed_return: if completed == 0 { 0.0 } else { minimum },
		maximum_completed_return: if completed == 0 { 0.0 } else { maximum },
	})
}

#[cfg(test)]
mod tests {
	use super::summarize;

	#[test]
	fn episode_summary_separates_lanes_and_both_boundary_kinds() -> crate::Result<()> {
		let metrics = summarize(
			3,
			2,
			&[1.0, 10.0, 2.0, 20.0, 4.0, 40.0],
			&[0, 0, 1, 0, 0, 0],
			&[0, 1, 0, 0, 1, 0],
		)?;
		assert_eq!(metrics.environment_steps, 3);
		assert_eq!(metrics.transitions, 6);
		assert_eq!(metrics.completed_episodes, 3);
		assert_eq!(metrics.minimum_completed_return, 3.0);
		assert_eq!(metrics.maximum_completed_return, 10.0);
		assert!((metrics.mean_completed_return - 17.0 / 3.0).abs() < f32::EPSILON);
		Ok(())
	}
}
